src/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -66,27 +66,35 @@ newtype IndexOrdinary = IndexOrdinary (Vector SerialisedKey)
     if there is no such pair, the result is an arbitrary but valid page span.
 -}
 search :: SerialisedKey -> IndexOrdinary -> PageSpan
-search key (IndexOrdinary lastKeys)
-    = assert (length lastKeys > 0) $ PageSpan (PageNo start) (PageNo end)
-    where
-
-    start :: Int
-    start | protoStart < pageCount = protoStart
-          | otherwise              = maybe 0 succ $
-                                     findIndexR (/= lastKeys ! pred pageCount) $
-                                     lastKeys
-
-    end :: Int
-    end | protoStart < pageCount
-            = start + length (takeWhile (== key) (drop (succ start) lastKeys))
-        | otherwise
-            = pred pageCount
+search key (IndexOrdinary lastKeys) = assert (length lastKeys > 0) result where
 
     protoStart :: Int
-    protoStart = binarySearchL lastKeys key
+    !protoStart = binarySearchL lastKeys key
 
     pageCount :: Int
-    pageCount = length lastKeys
+    !pageCount = length lastKeys
+
+    result :: PageSpan
+    !result | protoStart < pageCount
+                = let
+
+                      end :: Int
```

Comment:
I know it's a matter of taste, but the code feels very "airy" and spaced out, taking up a lot of screen space. Here for example, I'd personally not make these two line breaks.

Reply:
I know that I have the tendency of writing “airy” code. 🙂 Well, it’s also a matter of perspective: for example, I personally find some other code quite cramped.

As for the example you quoted: The definition of `end` is, well, a definition of its own, like the including definition of `result` and ultimately the definition of `search`. As such, it has some complexity, which might make it advisable to separate it from the surrounding code by one empty line on each side. For global definitions, this is what virtually everyone is doing, and doing it for local definitions as well has the further advantage of keeping the (implicit) style rules I’m following here simple and consistent.

As for saving screen space: This is the, quite usual, pragmatic argument. 🙂 However, I don’t think that these days there is a shortage of screen space. In fact, I believe that the typical person looking at Haskell code has much more screen space than I have, and I must say that I don’t suffer from looking at my code. 😉